x/gamm/keeper/msg_server_test.go

```diff
@@ -356,3 +361,160 @@ func (suite *KeeperTestSuite) TestExitPool_Events() {
+		clPool := suite.PrepareConcentratedPool()
+
+		// Note gamm and cl pool addresses
+		gammPoolAddress := gammPool.GetAddress()
+		clPoolAddress := clPool.GetAddress()
+
+		// Join gamm pool to create gamm shares directed in the test case
+		_, _, err = suite.App.GAMMKeeper.JoinPoolNoSwap(suite.Ctx, test.param.sender, gammPoolId, test.sharesToCreate, sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(999999999999999)), sdk.NewCoin("usdc", sdk.NewInt(999999999999999))))
+		suite.Require().NoError(err)
+
+		// Note gamm pool balance after joining gamm pool
+		gammPoolEthBalancePostJoin := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, ETH)
+		gammPoolUsdcBalancePostJoin := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, USDC)
+
+		// Create migrate message
+		sharesToMigrate := sdk.NewCoin(test.param.sharesToMigrateDenom, test.param.sharesToMigrateAmount)
+		msg := &balancer.MsgMigrateSharesToFullRangeConcentratedPosition{
+			Sender:          test.param.sender.String(),
+			SharesToMigrate: sharesToMigrate,
+			PoolIdEntering:  test.param.poolIdEntering,
+		}
+
+		// Reset event counts to 0 by creating a new manager.
+		suite.Ctx = suite.Ctx.WithEventManager(sdk.NewEventManager())
+		suite.Require().Equal(0, len(suite.Ctx.EventManager().Events()))
+
+		// Migrate the user's gamm shares to a full range concentrated liquidity position
+		resp, err := msgServer.MigrateSharesToFullRangeConcentratedPosition(sdk.WrapSDKContext(suite.Ctx), msg)
+		if test.expectedErr != nil {
+			suite.Require().Error(err)
+			suite.Require().ErrorContains(err, test.expectedErr.Error())
+			continue
+		}
+		suite.Require().NoError(err)
+
+		// Assert events are emitted
+		suite.AssertEventEmitted(suite.Ctx, types.TypeEvtMigrateShares, test.expectedMigrateShareEvents)
+		suite.AssertEventEmitted(suite.Ctx, sdk.EventTypeMessage, test.expectedMessageEvents)
+
+		// Note gamm pool balance after migration
+		gammPoolEthBalancePostMigrate := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, ETH)
+		gammPoolUsdcBalancePostMigrate := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, USDC)
+
+		// Note user amount transferred to cl pool from gamm pool
+		userEthBalanceTransferredToClPool := gammPoolEthBalancePostJoin.Sub(gammPoolEthBalancePostMigrate)
+		userUsdcBalanceTransferredToClPool := gammPoolUsdcBalancePostJoin.Sub(gammPoolUsdcBalancePostMigrate)
+
+		// Note cl pool balance after migration
+		clPoolEthBalanceAfterMigration := suite.App.BankKeeper.GetBalance(suite.Ctx, clPoolAddress, ETH)
+		clPoolUsdcBalanceAfterMigration := suite.App.BankKeeper.GetBalance(suite.Ctx, clPoolAddress, USDC)
+
+		// The balance in the cl pool should be equal to what the user previously had in the gamm pool
+		// This test is within 100 shares due to rounding that occurs from utilizing .000000000000000001 instead of 0
+		poolEthBalanceInTolerance := userEthBalanceTransferredToClPool.Amount.Sub(clPoolEthBalanceAfterMigration.Amount).LTE(sdk.NewInt(1))
```

Comment:
These checks should probably use `Compare`, and the tolerance should be extracted into a constant. If you only want to check tolerance in one direction, you can use the `RoundingDirection` field in `ErrTolerance` (this is how we check rounding behavior in stableswap)

Right now, it trivially passes if the result of the subtraction is negative

Reply:
Also it seems this check is comparing the user's transferred amount to the entire gamm pool's balance – this means it always passes trivially no?

Reply:
Addressed here https://github.com/osmosis-labs/osmosis/pull/4051/commits/65ed90e59267a06b1515671f21ffbae0cadb20ef

This check does not compare the user's transferred amount to the entire gamm pool balance though. It compares the cl pool balance to what was removed from the gamm pool after the migration, so it does not pass trivially